### PR TITLE
Publish robot statuses

### LINKED ISSUE #222: Handle Robot Statuses from Radio
**Is your feature request or enhancement related to a problem? What specifically could be better? Please describe.**
Statuses like cap voltages, battery voltage, temperatures, etc. are returned from the robot over radio. These statuses can be displayed in the visualiser and/or logs to aid in debugging robots.

**Describe the solution you'd like**
There are callbacks in the dongle class that handle robot status data. We can use the existing robot status handling code (in `mrf/robot.cpp`) to translate these into readable data, and then publish them from the radio_communication node.

**Describe alternatives you've considered**
N/A

**Additional context**

### PR DESCRIPTION
**Pull request by henryymliu**
### Description

Reads robot status over radio (voltages, statuses, etc) and publishes it to the node `backend/robot_status` (constant defined in `util/constants.h`)

~TODO: publish warning messages related to these statuses (ex. low battery, dead robot, etc.)~

@MathewMacDougall I changed the name of your ROBOT_STATUS_TOPIC for logging to ROBOT_MESSAGE_TOPIC if that's okay with you

### Testing Done

Tested with robot and `rostopic echo backend/robot_status`

### Resolved Issues

resolves #222 
### Length Justification

Legacy code? :/
### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/thunderbots/software/util/codec.cpp [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+    }
+    else if (!exponent)
+    {
+        // Exponent 0x000 encodes zeroes and subnormals.
+        // Shift the significand down into the fraction part.
+        double value = static_cast<double>(significand) / (UINT64_C(1) << 52);
+
+        // Apply the exponent (this value is 2^1022).
+        value /=
+            44942328371557897693232629769725618340449424473557664318357520289433168951375240783177119330601884005280028469967848339414697442203604155623211857659868531094441973356216371319075554900311523529863270738021251442209537670585615720368478277635206809290837627671146574559986811484619929076208839082406056034304.0;
```
Comment (henryymliu) [inline src/thunderbots/software/util/codec.cpp]: what the actual-

File: src/thunderbots/software/radio_communication/mrf/dongle.cpp [hunk 2]
```diff
@@ -288,25 +291,16 @@ void MRFDongle::handle_mdrs(AsyncOperation<void> &op)
...
-    //             robot, transfer.data() + 1, transfer.size() - 3,
-    //             transfer.data()[transfer.size() - 2],
-    //             transfer.data()[transfer.size() - 1]);
-    //     }
-    //     robots[robot]->handle_message(
-    //         transfer.data() + 1, transfer.size() - 3,
-    //         transfer.data()[transfer.size() - 2],
-    //         transfer.data()[transfer.size() - 1]);
-    // }
+    if (transfer.size() > 2)
```
Comment (garethellis0) [inline src/thunderbots/software/radio_communication/mrf/dongle.cpp:297]: What exactly are we checking here?

File: src/thunderbots/software/radio_communication/mrf_backend.cpp [hunk 3]
```diff
@@ -1,18 +1,100 @@
...
+        {233, -41}, {228, -42}, {225, -43}, {221, -44}, {216, -45}, {212, -46},
+        {207, -47}, {203, -48}, {198, -49}, {193, -50}, {188, -51}, {183, -52},
+        {176, -53}, {170, -54}, {165, -55}, {159, -56}, {153, -57}, {148, -58},
+        {143, -59}, {138, -60}, {133, -61}, {129, -62}, {125, -63}, {121, -64},
+        {117, -65}, {111, -66}, {107, -67}, {100, -68}, {95, -69},  {89, -70},
+        {83, -71},  {78, -72},  {73, -73},  {68, -74},  {63, -75},  {58, -76},
+        {53, -77},  {48, -78},  {43, -79},  {37, -80},  {32, -81},  {27, -82},
+        {23, -83},  {18, -84},  {13, -85},  {9, -86},   {5, -87},   {2, -88},
+        {1, -89},   {0, -90},
+    };
```
Comment (garethellis0) [inline src/thunderbots/software/radio_communication/mrf_backend.cpp]: What are there?

File: src/thunderbots/software/radio_communication/mrf_backend.cpp [hunk 4]
```diff
@@ -1,18 +1,100 @@
...
+     */
+    const unsigned int REQUEST_BUILD_IDS_COUNT = 7;
+
+    /**
+     * \brief The number of seconds to wait between consecutive requests for
+     * the build IDs.
+     */
+    const double REQUEST_BUILD_IDS_INTERVAL = 0.5;
+
+    struct RSSITableEntry final
```
Comment (garethellis0) [inline src/thunderbots/software/radio_communication/mrf_backend.cpp]: Quick javadoc function for this struct would be good.

File: src/thunderbots/software/radio_communication/mrf_backend.cpp [hunk 5]
```diff
@@ -1,18 +1,100 @@
...
+        nullptr,
+        u8"Bot %1 logger uninitialized",
+        nullptr,
+        u8"Bot %1 SD card full",
+    };
+}  // namespace
+
+MRFBackend::MRFBackend(ros::NodeHandle &node_handle)
+    : dongle(MRFDongle(std::bind(&MRFBackend::handle_message, this, ph::_1, ph::_2,
+                                 ph::_3, ph::_4, ph::_5))),
```
Comment (garethellis0) [inline src/thunderbots/software/radio_communication/mrf_backend.cpp]: scary bind is scary, let's add a quick comment roughly explaining what this does.....

File: src/thunderbots/software/radio_communication/mrf_backend.cpp [hunk 6]
```diff
@@ -50,3 +132,261 @@ void MRFBackend::update_dongle_events()
...
+                found                                    = true;
+            }
+        }
+        if (!found)
+        {
+            robot_status.received_signal_strength_db = -90;
+        }
+    }
+
+    const uint8_t *bptr = static_cast<const uint8_t *>(data);
```
Comment (garethellis0) [inline src/thunderbots/software/radio_communication/mrf_backend.cpp]: `bptr`?

File: src/thunderbots/software/radio_communication/mrf_backend.cpp [hunk 7]
```diff
@@ -50,3 +132,261 @@ void MRFBackend::update_dongle_events()
...
+                    robot_status.ball_in_beam      = !!(*bptr & 0x80);
+                    robot_status.capacitor_charged = !!(*bptr & 0x40);
+                    // unsigned int logger_status = *bptr & 0x3F;
+                    // for (std::size_t i = 0; i < logger_messages.size(); ++i)
+                    // {
+                    //     if (logger_messages[i])
+                    //     {
+                    //         logger_messages[i]->active(logger_status == i);
+                    //     }
+                    // }
```
Comment (garethellis0) [inline src/thunderbots/software/radio_communication/mrf_backend.cpp]: Can we remove this?

File: src/thunderbots/software/radio_communication/mrf_backend.cpp [hunk 8]
```diff
@@ -50,3 +132,261 @@ void MRFBackend::update_dongle_events()
...
+                    ++bptr;
+                    --len;
+
+                    // for (std::size_t i = 0; i < sd_messages.size(); ++i)
+                    // {
+                    //     if (sd_messages[i])
+                    //     {
+                    //         sd_messages[i]->active(*bptr == i);
+                    //     }
+                    // }
```
Comment (garethellis0) [inline src/thunderbots/software/radio_communication/mrf_backend.cpp]: Can we remove this?

File: src/thunderbots/software/radio_communication/mrf_backend.cpp [hunk 9]
```diff
@@ -50,3 +132,261 @@ void MRFBackend::update_dongle_events()
...
+                                    has_error_extension = true;
+                                    for (unsigned int i = 0; i != MRF::ERROR_LT_COUNT;
+                                         ++i)
+                                    {
+                                        unsigned int byte = i / CHAR_BIT;
+                                        unsigned int bit  = i % CHAR_BIT;
+
+                                        // TODO: Handle these messages
+                                        // error_lt_messages[i]->active(
+                                        //     bptr[byte] & (1 << bit));
```
Comment (garethellis0) [inline src/thunderbots/software/radio_communication/mrf_backend.cpp]: Can we remove this?

File: src/thunderbots/software/radio_communication/mrf_backend.cpp [hunk 10]
```diff
@@ -50,3 +132,261 @@ void MRFBackend::update_dongle_events()
...
+
+                            case 0x01:  // Build IDs.
+                                ++bptr;
+                                --len;
+                                if (len >= 8)
+                                {
+                                    robot_status.build_ids_valid = true;
+                                    robot_status.fw_build_id     = decode_u32_le(bptr);
+                                    robot_status.fpga_build_id = decode_u32_le(bptr + 4);
+                                    // TODO: check this
```
Comment (garethellis0) [inline src/thunderbots/software/radio_communication/mrf_backend.cpp]: ?

File: src/thunderbots/software/radio_communication/mrf_backend.cpp [hunk 11]
```diff
@@ -50,3 +132,261 @@ void MRFBackend::update_dongle_events()
...
+                            case 0x02:  // LPS data.
+                                ++bptr;
+                                --len;
+                                if (len >= 4)
+                                {
+                                    // for (unsigned int i = 0; i < 4; ++i)
+                                    // {
+                                    //     lps_values[i] =
+                                    //         static_cast<int8_t>(*bptr++) / 10.0;
+                                    // }
```
Comment (garethellis0) [inline src/thunderbots/software/radio_communication/mrf_backend.cpp]: Can we remove this?

File: src/thunderbots/software/radio_communication/mrf_backend.h [hunk 12]
```diff
@@ -49,8 +54,23 @@ class MRFBackend
...
+     * 
+     * @param index robot number
+     * @param data the message data
+     * @param len the length of the message
+     * @param lqi link quality
+     * @param rssi received signal strength indicator
+     */
+    thunderbots_msgs::RobotStatus handle_message(int index, const void* data,
+                                                 std::size_t len, uint8_t lqi,
+                                                 uint8_t rssi);
```
Comment (garethellis0) [inline src/thunderbots/software/radio_communication/mrf_backend.h]: This is a clean interface!

File: src/thunderbots/software/util/codec.cpp [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+    {
+        // Extract sign bit and absolutize input.
+        bool sign = !!std::signbit(x);
+        if (sign)
+        {
+            x = -x;
+        }
+
+// Extract exponent.
+#if FLT_RADIX == 2
```
Comment (garethellis0) [inline src/thunderbots/software/util/codec.cpp]: nani!?!

File: src/thunderbots/software/util/codec.cpp (hunk 1 above)
Comment (garethellis0) [reply src/thunderbots/software/util/codec.cpp]: really tho. Perhaps `const some_var_name = std::pow(2,1022)` _might_ be a bit better........?

File: src/thunderbots_msgs/msg/RobotStatus.msg [hunk 14]
```diff
@@ -0,0 +1,82 @@
+
+# The rough maximum full-scale deflection of the laser sensor.
+float64 break_beam_scale
```
Comment (garethellis0) [inline src/thunderbots_msgs/msg/RobotStatus.msg]: deflection? scale? what do these mean?

File: src/thunderbots_msgs/msg/RobotStatus.msg [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+
+# The speed of the robot’s dribbler motor, in revolutions per minute.
+int32 dribbler_speed
+
+
+# The temperature of the robot’s mainboard, in degrees Celsius.
+float64 board_temperature
+
+
+# The link quality of the last received packet, from 0 to 1.
```
Comment (garethellis0) [inline src/thunderbots_msgs/msg/RobotStatus.msg]: Is 1 good? or is 0 good? (I'm guessing 1, but might as well be super clear here)

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Most the comments are just about adding some doc comments and whatnot so we roughly know what's going on. My other question would be, do we plan on modifying or working with this at all before we toss radio entirely? @MathewMacDougall ?

In general I can't even begin to understand all of this, so I'm leaving most of the nasty radio bits untouched and taking a more detailed look at the slightly less.... _creative_ bits that we're actually going to interact with. I'll try and take another pass through at some point, just because there's so much here.

- I pointed out a few places, but we should remove all commented out code or assign it to an issue

File: src/thunderbots/software/util/codec.cpp (hunk 1 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/util/codec.cpp]: Holy...

**Comment by MathewMacDougall** [comment]
@garethellis0 I don't expect us to need to change anything with how robot statuses are reported, so this will hopefully not be an issue. It's much more likely we will change the encoding for how primitives are sent. This code makes me _really_ want WiFi to come through...

@henrymliu :+1: re: topic name change

Not much to add on Gareth's comments right now, will do another pass later.

**Comment by henryymliu** [comment]
This is still pretty far from being done, so those comments will be addressed eventually.
The code ported over doesn't exactly give me joy, might have to rewrite that entire function once I start to send messages over in addition to the other values.

File: src/thunderbots/software/radio_communication/mrf_backend.h [hunk 16]
```diff
@@ -49,8 +54,23 @@ class MRFBackend
      */
     void send_vision_packet();
 
+    /**
+     * Parses a robot status message from the dongle, and publishes it to the ROS topic.
+     * 
+     * @param index robot number
+     * @param data the message data
+     * @param len the length of the message
+     * @param lqi link quality
```
Comment (jordanzliu) [inline src/thunderbots/software/radio_communication/mrf_backend.h]: To clarify, does LQI work like signal bars on a phone, or does it have a unit?

Comment (henryymliu) [reply src/thunderbots/software/radio_communication/mrf_backend.h]: Oops, sorry for the late reply, I think RSSI is more related to signal bar strength? I can guess link quality is a function of RSSI, packet drop rate (maybe?), and others.

File: src/thunderbots_msgs/msg/RobotStatus.msg (hunk 14 above)
Comment (henryymliu) [reply src/thunderbots_msgs/msg/RobotStatus.msg]: unfortunately, I have no idea

**Comment by henryymliu** [comment]
@MathewMacDougall @garethellis0 ready for proper review
Added @qgolsteyn to review RobotStatus.msg, let me know if the contents are reasonable

File: src/thunderbots/software/util/codec.cpp (hunk 1 above)
Comment (henryymliu) [reply src/thunderbots/software/util/codec.cpp]: removed the file, this is not an issue anymore

File: src/thunderbots/software/radio_communication/mrf_backend.cpp (hunk 3 above)
Comment (henryymliu) [reply src/thunderbots/software/radio_communication/mrf_backend.cpp]: left a comment

File: src/thunderbots/software/radio_communication/mrf_backend.cpp (hunk 5 above)
Comment (henryymliu) [reply src/thunderbots/software/radio_communication/mrf_backend.cpp]: removed, not relevant anymore

File: src/thunderbots/software/radio_communication/mrf_backend.cpp (hunk 6 above)
Comment (henryymliu) [reply src/thunderbots/software/radio_communication/mrf_backend.cpp]: ...don't know of any better name, that was part of the legacy code

File: src/thunderbots/software/util/codec.cpp (hunk 13 above)
Comment (henryymliu) [reply src/thunderbots/software/util/codec.cpp]: file removed

File: src/thunderbots_msgs/msg/RobotStatus.msg (hunk 15 above)
Comment (henryymliu) [reply src/thunderbots_msgs/msg/RobotStatus.msg]: left a comment

File: src/thunderbots/software/radio_communication/mrf/dongle.cpp (hunk 2 above)
Comment (henryymliu) [reply src/thunderbots/software/radio_communication/mrf/dongle.cpp:297]: left a comment

File: src/thunderbots/software/util/codec.cpp (hunk 1 above)
Comment (garethellis0) [reply src/thunderbots/software/util/codec.cpp]: well that's one way to solve the problem..............

File: src/thunderbots/software/radio_communication/mrf_backend.cpp (hunk 5 above)
Comment (garethellis0) [reply src/thunderbots/software/radio_communication/mrf_backend.cpp]: my fav. soln

File: src/thunderbots_msgs/msg/RobotStatus.msg (hunk 14 above)
Comment (garethellis0) [reply src/thunderbots_msgs/msg/RobotStatus.msg]: Ah, I believe that this is the estimated maximum value for the break beam reading (the "deflection" part comes from the needle deflection on an older analog reading: https://electronics.stackexchange.com/questions/297210/what-is-full-scale-deflection). Could we move this down with `break_beam_reading` and add a quick comment this effect so it's clear?

Comment (henryymliu) [reply src/thunderbots_msgs/msg/RobotStatus.msg]: done :+1:

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada: - great work getting this guy in!

**Review by MathewMacDougall — APPROVED** [review]
LGTM :+1: :tada: You're a hero.

**Comment by garethellis0** [comment]
> LGTM +1 tada You're a hero.

https://www.youtube.com/watch?v=iovcnjuf2WU

**Review by qgolsteyn — APPROVED** [review]
Sorry for taking so long. LGTM :+1:

**Pull request merged**